src/eos-updater-apply.c

```diff
@@ -73,6 +74,431 @@ get_test_osname (void)
+          !g_hash_table_contains (already_installed_flatpaks,
+                                  flatpak_to_install->app_identifier))
+        {
+          g_ptr_array_add (resolved_flatpaks,
+                           flatpak_to_install_copy (flatpak_to_install));
+        }
+    }
+
+  return g_steal_pointer (&resolved_flatpaks);
+}
+
+static FlatpakInstallation *
+get_flatpak_installation (GCancellable *cancellable, GError **error)
+{
+  const gchar *override_path = g_getenv ("EOS_UPDATER_TEST_FLATPAK_INSTALLATION_DIR");
+
+  if (override_path)
+    {
+      g_autoptr(GFile) override = g_file_new_for_path (override_path);
+      return flatpak_installation_new_for_path (override, TRUE, cancellable, error);
+    }
+
+  return flatpak_installation_new_system (cancellable, error);
+}
+
+static gboolean
+pull_flatpaks (GPtrArray    *flatpaks_to_deploy,
+               GCancellable *cancellable,
+               GError       **error)
+{
+  g_autoptr(FlatpakInstallation) installation = get_flatpak_installation (cancellable,
+                                                                          error);
+  gsize i = 0;
+
+  if (!installation)
+    return FALSE;
+
+  for (; i < flatpaks_to_deploy->len; ++i)
+    {
+      FlatpakToInstall *to_install = g_ptr_array_index (flatpaks_to_deploy, i);
+      g_autoptr(GError) local_error = NULL;
+
+      /* We have to pass in a local_error instance here and check to see
+       * if it was FLATPAK_ERROR_NOT_INSTALLED - this is what will be
+       * thrown if we succeeded at pulling the flatpak into the local
+       * repository but did not deploy it (since no FlatpakInstalledRef
+       * will be returned). */
+      flatpak_installation_install_full (installation,
+                                         FLATPAK_INSTALL_FLAGS_NO_DEPLOY,
+                                         to_install->repo,
+                                         to_install->ref_kind,
+                                         to_install->app_identifier,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         cancellable,
+                                         &local_error);
+
+      /* This only happens if the wrong version of flatpak is deployed
+       * with eos-updater */
+      if (!local_error)
+        {
+          g_critical ("Flatpak installation should not have succeeded!");
+          g_assert (FALSE);
+        }
+
+      /* Something unexpected failed, return early now.
+       *
+       * We are not able to meaningfully clean up here - the refs will remain
+       * in the flatpak ostree repo for the next time we want to install them
+       * but there's nothing in the public API that will allow us to get rid of
+       * them. */
```

Comment:
Can we add anything to the public API (or otherwise improve libflatpak) to help here?

Reply:
We *could*, but I was thinking just now that maybe we can bypass all this by using ostree on the underlying repo to get rid of the refs directly.

Reply:
> maybe we can bypass all this by using ostree on the underlying repo to get rid of the refs directly.

We *could* use libostree to do that, but that runs the risk of pulling the rug out from underneath libflatpak if our code drifts out of sync with how libflatpak treats refs in future. I’d rather do it within libflatpak.

Reply:
I'll file a ticket upstream. The only thing I'm a little weary about is that we are already in niche-feature territory and increasing the API surface of flatpak doesn't intuitively feel right to me. But I'll ask to be sure anyway.